AVRO-3375: Better error messages out of generic and specific DatumWriters

radai-rosenblatt opened the pull request:
this commit introduces "path tracking" to trace the path from a root datum object to be written to the location of 3 common serialization errors: NullPointerExceptions, ClassCastExceptions and AvroTypeExceptions.
    
path tracing is done using AvroPath-like path components and predicates that would maybe become part of a future implementation of AVRO-992
    
note that path tracing does not work if custom coders are in use for serialization (neither does the tracking code this commit replaces).

before/after exception messages, by errror

- generic NPE
   - null of string in field f4 of org.apache.avro.MapRecord of map in field mapField of org.apache.avro.ArrayRecord of array in field arrayField of org.apache.avro.UnionRecord of union in field unionField of org.apache.avro.RecordWithRequiredFields
   - null value for (non-nullable) string at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField("a").f4
- generic NPE in map key
   - null of map in field mapField of org.apache.avro.ArrayRecord of array in field arrayField of org.apache.avro.UnionRecord of union in field unionField of org.apache.avro.RecordWithRequiredFields
   - null key in map at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField
- generic ClassCastException
   - java.lang.Integer cannot be cast to java.lang.CharSequence in field f4 in field mapField in field arrayField in field unionField
   - value 42 (a java.lang.Integer) cannot be cast to expected type string at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField("a").f4
- generic AvroTypeException
   - Not an enum: 42 for schema: {"type":"enum","name":"MapRecordEnum","namespace":"org.apache.avro","symbols":["A","B"],"default":"A"} in field f3 in field mapField in field arrayField in field unionField
   - value 42 (a java.lang.Integer) is not a MapRecordEnum at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField("a").f3
- specific NPE
   - null of string in field 'f4' of org.apache.avro.MapRecord of map in field 'mapField' of org.apache.avro.ArrayRecord of array in field 'arrayField' of org.apache.avro.UnionRecord of union in field 'unionField' of 
   - org.apache.avro.RecordWithRequiredFields
null value for (non-nullable) string at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField("a").f4
- specific NPE in map key
   - null of map in field 'mapField' of org.apache.avro.ArrayRecord of array in field 'arrayField' of org.apache.avro.UnionRecord of union in field 'unionField' of org.apache.avro.RecordWithRequiredFields
   - null key in map at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField
- specific AvroTypeException
   - Not an enum: null for schema: {"type":"enum","name":"MapRecordEnum","namespace":"org.apache.avro","symbols":["A","B"],"default":"A"} in field 'f3' in field 'mapField' in field 'arrayField' in field 'unionField'
   - value null is not a MapRecordEnum at RecordWithRequiredFields.unionField[UnionRecord].arrayField[0].mapField("a").f3

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3375
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests added in TestGenericDatumWriter and TestSpecificDatumWriter

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

martin-g commented on lang/java/avro/src/main/java/org/apache/avro/generic/TracingNullPointException.java:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  @Override
+  public synchronized NullPointerException getCause() {
```
why `synchronized` ?

martin-g commented on lang/java/avro/src/main/java/org/apache/avro/generic/TracingNullPointException.java:
```diff
@@ -0,0 +1,74 @@
+   */
+  public NullPointerException summarize() {
+    StringBuilder sb = new StringBuilder();
+    PathStep lastStep = reversePath.get(0);
```
it looks possible this may throw IndexOutOfBoundException

radai-rosenblatt commented:
@martin-g - addressed your comments (and also handled the case of a null map key, which i dont think avro currently handles correctly).

im going to replicate this work for other exception types and writer subclasses

radai-rosenblatt commented:
@RyanSkraba @martin-g - Code has been cleaned up and the "path steps" have been converted over to something that might one day form part of an AvroPath implementation.

i'll send out an email to the dev mailing list later today about formalizing at least enough syntax for tracking the locations of errors in graphs, but otherrwise this PR is ready for review

martin-g commented on share/test/schemas/RecordWithRequiredFields.avsc:
```diff
@@ -0,0 +1,53 @@
+  "name" : "RecordWithRequiredFields",
+  "namespace" : "org.apache.avro",
+  "fields" : [ {
+    "name" : "f0",
```
IMO it would be easier to understand the test cases if `fN` fields were named after their types, e.g. `stringField` and `enumField`.
Since this PR makes it easier to pinpoint the problematic field I see no problem in using `stringField` several types in the schema.

radai-rosenblatt replied:
will do

radai-rosenblatt commented on lang/java/avro/src/main/java/org/apache/avro/PathTracingException.java:
```diff
@@ -0,0 +1,39 @@
+ * limitations under the License.
+ */
+
+package org.apache.avro;
```
will do

radai-rosenblatt commented on lang/java/avro/src/main/java/org/apache/avro/generic/TracingClassCastException.java:
```diff
@@ -0,0 +1,85 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.avro.generic;
```
i think simply because i initially solved for generic records. i will be moving all tracking* classes to the path package

radai-rosenblatt commented on lang/java/avro/src/main/java/org/apache/avro/path/PathPredicate.java:
```diff
@@ -0,0 +1,22 @@
+
+package org.apache.avro.path;
+
+public interface PathPredicate extends PathElement {
```
jspath calls these "location paths" - https://github.com/dfilatov/jspath#location-path. i'll rename to something similar. LocationElement perhaps?

martin-g replied:
PathElement is clear.
I don't understand the purpose of the specialization (PathPredicate, or any other name for it).

radai-rosenblatt replied:
i've renamed things to more closely follow jspath. i've also added javadocs to all interfaces and classes. hopefully its clearer now?

radai-rosenblatt replied:
do you maybe mean why i added all the intermediate interfaces with no content? 
reason for that is that if a full blown AvroPath implementation does follow these interfaces will likely be required.
I can remove all the intermediates if thats the concern

martin-g replied:
Let's see what others will say.

radai-rosenblatt commented:
addressed review comments.
also changed MapKeyPredicate to use square brackets like the others for consistency

RyanSkraba reviewed (approved):
Thanks @martin-g for your thorough review and @radai-rosenblatt for this contribution!

I think this is a great starting point because meaningful error messages are (1) highly visible and (2) very useful.   I don't expect this to be perfect, but I would not be surprised if this helps drive what we could do with an "avropath" selector standard!

If we had an `@Experimental` tag, I would ask you to put it on the classes in `org.apache.avro.path` ... Maybe just a doc in a package.html?  To let people know that these shouldn't be considered API classes, but that we are actively soliciting feedback based on this first implementation?

I think this should go in the next minor release for that reason!

Also, sorry for the bother -- I think the `synchronized` keywords were actually correct, even if it's not super important, it'll clean up the code scanning warnings.

radai-rosenblatt commented:
synchronized is back, package-info.java and package.html added

radai-rosenblatt commented:
build failure is a connection timeout downloading https://repo.maven.apache.org/maven2/commons-cli/commons-cli/1.5.0/commons-cli-1.5.0.pom

RyanSkraba commented on lang/java/avro/src/main/java/org/apache/avro/path/package-info.java:
```diff
@@ -0,0 +1,25 @@
+ */
+
+/**
+ * Interfaces and base classes for AvroPath. This functionality is
```
Perfect!

radai-rosenblatt commented:
@RyanSkraba i see master is already 1.12 - can i file a backport PR for this into 1.11 after this one is merged to master?

martin-g commented:
It is a new feature, so it should be only in 1.12.
Maybe we should release both 1.11.1 and 1.12.0 ?! But it is more work for the release manager.

radai-rosenblatt commented:
i'd appreciate a 1.11.1, but i'll take what i can get

martin-g commented:
Well, I am not sure. There are also other new features in branch-1.11, especially in the Rust SDK.

RyanSkraba commented:
Hmm I don't consider error messages and strings to be part of the API or a breaking change!  I vote for cherry-picking to 1.11.1 (remember that it's not semver... as awkward as it is, we've always consider these to be minor and not patch releases).

RyanSkraba commented:
Cherry-picked to [branch-1.11](https://github.com/apache/avro/commit/b0fbd92ddce1f44bd7b8b41f14069ec2e360087b).
